gnomad_qc/v4/create_release/calculate_variant_statistics.py

```diff
@@ -0,0 +1,416 @@
+    :param vep_mane: If `by_csqs` is True, filter to only MANE transcripts. Default is
+        False.
+    :param rare_variants_af: The allele frequency threshold to use for rare variants.
+        Default is 0.001.
+    :return: Table containing per-sample variant counts.
+    """
+    logger.info("Filtering input MT to variants in the supplied annotation HT...")
+    mt = mt.semi_join_rows(annotation_ht)
+
+    # Add extra Allele Count and Allele Type annotations to variant MatrixTable,
+    # according to Hail standards, to help their computation.
+    mt = mt.annotate_rows(
+        **vmt_sample_qc_variant_annotations(global_gt=mt.GT, alleles=mt.alleles)
+    )
+
+    keep_annotations = ["freq", "filters", "region_flags"]
+    if by_csqs:
+        annotation_ht = filter_vep_transcript_csqs(
+            annotation_ht,
+            synonymous=False,
+            canonical=vep_canonical,
+            mane_select=vep_mane,
+        )
+        annotation_ht = get_most_severe_consequence_for_summary(annotation_ht)
+        keep_annotations.extend(["most_severe_csq", "lof"])
+
+    # Annotate the MT with the needed annotations.
+    annotation_ht = annotation_ht.select(*keep_annotations).checkpoint(
+        hl.utils.new_temp_file("annotation_ht", "ht")
+    )
+    mt = mt.annotate_rows(**annotation_ht[mt.row_key])
+
+    filter_expr = {"all_variants": True}
+
+    if pass_filters:
+        filter_expr["pass_filters"] = hl.len(mt.filters) == 0
+    if ukb_capture:
+        filter_expr["ukb_capture"] = ~mt.region_flags.outside_ukb_capture_region
+    if broad_capture:
+        filter_expr["broad_capture"] = ~mt.region_flags.outside_broad_capture_region
+    if ukb_capture and broad_capture:
+        filter_expr["ukb_broad_capture_intersect"] = (
+            filter_expr["ukb_capture"] & filter_expr["broad_capture"]
+        )
+        filter_expr["ukb_broad_capture_union"] = (
+            filter_expr["ukb_capture"] | filter_expr["broad_capture"]
+        )
+    if rare_variants:
+        filter_expr["rare_variants"] = mt.freq[0].AF < rare_variants_af
+    if by_csqs:
+        filter_expr["lof"] = ~hl.is_missing(mt.lof)
+        filter_expr["missense"] = mt.most_severe_csq == "missense_variant"
+        filter_expr["synonymous"] = mt.most_severe_csq == "synonymous_variant"
+
+    # Run Hail's 'vmt_sample_qc' for all requested filter groups.
+    ht = mt.select_cols(
+        _sample_qc=hl.struct(
+            **{
+                ann: hl.agg.filter(
+                    expr,
+                    vmt_sample_qc(
+                        global_gt=mt.GT,
+                        gq=mt.GQ,
+                        variant_ac=mt.variant_ac,
+                        variant_atypes=mt.variant_atypes,
+                        dp=mt.DP,
+                    ),
+                )
+                for ann, expr in filter_expr.items()
+            }
+        )
+    ).cols()
+
+    return ht.select(**ht._sample_qc)
+
+
+def compute_agg_sample_stats(
```

Comment:
@KoalaQin This is what I was thinking about when I told you I had a plan for what I want this second part of the script to look like, but I haven't incorporated it in yet 

```
all_strats = [s for s in ht.row_value if isinstance(ht[s], hl.expr.StructExpression)]
ht = ht.transmute(
    gen_anc=["all_samples", meta_ht[ht.s].population_inference.pop],
    _stats_array=[(s, ht[s]) for s in all_strats]
)
ht = ht.explode("_stats_array").explode("gen_anc")
ht = ht.group_by("gen_anc", variant_filter=ht._stats_array[0]).aggregate(
    **{
        m: hl.struct(
            mean=hl.agg.mean(ht._stats_array[1][m]),
            quantiles=hl.agg.approx_quantiles(
                ht._stats_array[1][m] , [0.0, 0.25, 0.5, 0.75, 1.0]
            ),
        )
        for m in ht._stats_array[1] 
        if isinstance(ht._stats_array[1][m], hl.expr.NumericExpression)
    }
)
```

it still needs to support the by_ancestry and change the return type, and be written out to a new HT instead of added as globals